File: code/c00_shared/geography.R
# Canonical geographic identifiers used by the supported pipeline.
#
# Identifiers are character strings so leading zeroes survive CSV, Parquet,
# joins, and round trips through Python. County identifiers use the project's
# 2010 geography throughout the analysis pipeline.

clean_geo_code <- function(x) {
  x <- trimws(as.character(x))
  x <- gsub('"', "", x, fixed = TRUE)
  x <- sub("\\.0+$", "", x)
  x[x == ""] <- NA_character_
  x
}

normalize_fixed_width_code <- function(x, width, label) {
  x <- clean_geo_code(x)
  invalid <- !is.na(x) & (!grepl("^[0-9]+$", x) | nchar(x) > width)

  if (any(invalid)) {
    examples <- paste(utils::head(unique(x[invalid]), 5L), collapse = ", ")
    stop(
      label,
      " must contain at most ",
      width,
      " digits; invalid values: ",
      examples,
      call. = FALSE
    )
  }

  padding <- pmax(width - nchar(x), 0L)
  padded <- paste0(
    vapply(padding, function(n) strrep("0", n), character(1)),
    x
  )
  padded[is.na(x)] <- NA_character_
  padded
}

normalize_unpadded_code <- function(x, label) {
  x <- clean_geo_code(x)
  invalid <- !is.na(x) & !grepl("^[0-9]+$", x)

  if (any(invalid)) {
    examples <- paste(utils::head(unique(x[invalid]), 5L), collapse = ", ")
    stop(label, " must contain digits only; invalid values: ", examples,
      call. = FALSE
    )
  }

  x <- sub("^0+(?=[0-9])", "", x, perl = TRUE)
  x
}

state_fips <- function(x) {
  normalize_fixed_width_code(x, 2L, "state_fips")
}

county_code <- function(x) {
  normalize_fixed_width_code(x, 3L, "county_code")
}

county_fips <- function(x) {
  normalize_fixed_width_code(x, 5L, "county_fips")
}

combine_county_fips <- function(state, county) {
  paste0(state_fips(state), county_code(county))
}

state_from_county_fips <- function(county) {
  substr(harmonize_county_fips_2010(county), 1L, 2L)
}

county_code_from_county_fips <- function(county) {
  substr(harmonize_county_fips_2010(county), 3L, 5L)
}

harmonize_county_fips_2010 <- function(x) {
  x <- county_fips(x)
  x[x == "46102"] <- "46113"
  x
}

cz_id <- function(x) {
  normalize_unpadded_code(x, "cz_id")
}

aewr_region_id <- function(x) {
  x <- normalize_unpadded_code(x, "aewr_region_id")
  invalid <- !is.na(x) & !x %in% as.character(seq_len(17L))

  if (any(invalid)) {
    examples <- paste(utils::head(unique(x[invalid]), 5L), collapse = ", ")
    stop(
      "aewr_region_id must be between 1 and 17; invalid values: ",
      examples,
      call. = FALSE
    )
  }

  x
}

oews_area_code <- function(x) {
  x <- clean_geo_code(x)
  invalid <- !is.na(x) & !grepl("^[0-9]+$", x)

  if (any(invalid)) {
    examples <- paste(utils::head(unique(x[invalid]), 5L), collapse = ", ")
    stop(
      "oews_area_code must contain digits only; invalid values: ",
      examples,
      call. = FALSE
    )
  }

  x
}

geo_code_is_valid <- function(x, name) {
  if (!is.character(x)) {
    return(rep(FALSE, length(x)))
  }

  switch(
    name,
    state_fips = is.na(x) | grepl("^[0-9]{2}$", x),
    county_code = is.na(x) | grepl("^[0-9]{3}$", x),
    county_fips = is.na(x) | (grepl("^[0-9]{5}$", x) & x != "46102"),
    neighbor_county_fips = is.na(x) |
      (grepl("^[0-9]{5}$", x) & x != "46102"),
    cz_id = is.na(x) | grepl("^(0|[1-9][0-9]*)$", x),
    aewr_region_id = is.na(x) | x %in% as.character(seq_len(17L)),
    oews_area_code = is.na(x) | grepl("^[0-9]+$", x),
    stop("Unknown geographic identifier: ", name, call. = FALSE)
  )
}

assert_geo_columns <- function(data, required, allow_na = character()) {
  missing_columns <- setdiff(required, names(data))
  if (length(missing_columns) > 0L) {
    stop(
      "Missing required geographic columns: ",
      paste(missing_columns, collapse = ", "),
      call. = FALSE
    )
  }

  for (name in required) {
    value <- data[[name]]

    if (!is.character(value)) {
      stop(name, " must be a character vector.", call. = FALSE)
    }

    invalid <- !geo_code_is_valid(value, name)
    if (any(invalid)) {
      examples <- paste(
        utils::head(unique(value[invalid]), 5L),
        collapse = ", "
      )
      stop(name, " contains malformed values: ", examples, call. = FALSE)
    }

    if (!name %in% allow_na && anyNA(value)) {
      stop(name, " contains missing values.", call. = FALSE)
    }
  }

  invisible(data)
}

File: src/h2a/geography.py
"""Canonical geographic identifiers for Python-owned pipeline artifacts."""

from __future__ import annotations

from collections.abc import Iterable
from typing import Any

import polars as pl

# docs-ground:start geographic-code-contract-python
_FIXED_WIDTH = {
    "state_fips": 2,
    "county_code": 3,
    "county_fips": 5,
    "neighbor_county_fips": 5,
}
_UNPADDED = {"cz_id", "aewr_region_id"}
_VARIABLE_WIDTH = {"oews_area_code"}


def clean_geo_code(value: Any) -> str | None:
    """Clean one source value without changing its substantive digits."""
    if value is None:
        return None
    text = str(value).strip().replace('"', "")
    if text.endswith(".0") and text[:-2].isdigit():
        text = text[:-2]
    return text or None


def normalize_geo_code(value: Any, name: str) -> str | None:
    """Return one canonical identifier or raise for a malformed value."""
    text = clean_geo_code(value)
    if text is None:
        return None
    if not text.isdigit():
        raise ValueError(f"{name} must contain digits only: {value!r}")

    if name in _FIXED_WIDTH:
        width = _FIXED_WIDTH[name]
        if len(text) > width:
            raise ValueError(f"{name} must contain at most {width} digits: {value!r}")
        text = text.zfill(width)
    elif name in _UNPADDED:
        text = text.lstrip("0") or "0"
    elif name in _VARIABLE_WIDTH:
        pass
    else:
        raise ValueError(f"Unknown geographic identifier: {name}")

    if name == "aewr_region_id" and text not in {str(i) for i in range(1, 18)}:
        raise ValueError(f"aewr_region_id must be between 1 and 17: {value!r}")
    return text


def harmonize_county_fips_2010(value: Any) -> str | None:
    """Normalize a county identifier to the project's 2010 county vintage."""
    county = normalize_geo_code(value, "county_fips")
    return "46113" if county == "46102" else county


def state_from_county_fips(value: Any) -> str | None:
    """Return the canonical state component of a county FIPS."""
    county = harmonize_county_fips_2010(value)
    return None if county is None else county[:2]


def county_code_from_county_fips(value: Any) -> str | None:
    """Return the canonical three-digit component of a county FIPS."""
    county = harmonize_county_fips_2010(value)
    return None if county is None else county[2:]


def normalize_county_fips_list(value: Any) -> str:
    """Normalize a comma-delimited set of candidate counties."""
    text = clean_geo_code(value)
    if text is None:
        return ""
    counties = [
        harmonize_county_fips_2010(part)
        for part in text.split(",")
        if part.strip()
    ]
    return ",".join(county for county in counties if county is not None)


def geo_expr(source: str, name: str, *, vintage_2010: bool = False) -> pl.Expr:
    """Build a Polars expression that normalizes a geographic column."""
    normalizer = (
        harmonize_county_fips_2010
        if vintage_2010
        else lambda value: normalize_geo_code(value, name)
    )
    return (
        pl.col(source)
        .map_elements(normalizer, return_dtype=pl.String)
        .alias(name)
    )


def assert_geo_columns(
    frame: pl.DataFrame,
    required: Iterable[str],
    *,
    allow_null: Iterable[str] = (),
) -> None:
    """Fail when an artifact does not satisfy the canonical geo contract."""
    required = tuple(required)
    allow_null = set(allow_null)
    missing = [name for name in required if name not in frame.columns]
    if missing:
        raise ValueError(f"Missing required geographic columns: {', '.join(missing)}")

    for name in required:
        if frame.schema[name] != pl.String:
            raise TypeError(f"{name} must use Polars String, got {frame.schema[name]}")
        values = frame.get_column(name)
        if name not in allow_null and values.null_count() > 0:
            raise ValueError(f"{name} contains missing values")
        for value in values.drop_nulls().unique().to_list():
            normalized = (
                harmonize_county_fips_2010(value)
                if name in {"county_fips", "neighbor_county_fips"}
                else normalize_geo_code(value, name)
            )
            if normalized != value:
                raise ValueError(f"{name} contains noncanonical value {value!r}")
# docs-ground:end geographic-code-contract-python
